### ISSUE TITLE
add tests into the workflow

### ISSUE THREAD

**Issue by samueleternity** [samueleternity/initium#26]

**Comment by samueleternity** [comment]
Unit tests, Integration tests (Controllers/models work together), Smoke test (Model can actually instantiate/run).

**Comment by samueleternity** [comment]
Create tests around the mathematically meaningful / stateful pieces. Gradient tests. Checkpoint round-trip. Tiny end-to-end smoke test

**Comment by samueleternity** [comment]
GitHub Actions should run:

1. formatting check
2. lint
3. type check
4. unit tests
5. model/controller smoke tests
6. optionally a small CPU integration test

**Comment by samueleternity** [comment]
model:
    hidden_size = 16
    memory_size = 16
    1–2 layers
    tiny sequence length

dataset:
    synthetic
    maybe 32–128 samples

training:
    2–10 optimizer steps

**Comment by samueleternity** [comment]
Test every (in meaningful way, something like mamba+cfc is enough and testing cfc+mamba wouldnt make much sense because effectively both would launch if one of them does) controller type (plain controller or split-graph) in very small training cycle

**Comment by samueleternity** [comment]
No 100% coverage needed

**Comment by samueleternity** [comment]
use pytest + PyTorch's own testing utilities

**Comment by samueleternity** [comment]
pytest
  │
  ├── fixtures
  ├── parametrized tests
  ├── assertions
  ├── test discovery
  └── CI integration
       +
PyTorch testing utilities
  │
  ├── torch.testing.assert_close()
  └── gradient/numerical testing utilities
       +
own test cases
  │
  ├── model shapes
  ├── routing correctness
  ├── gradients
  ├── checkpoint round-trips
  └── tiny training runs
